### DISCUSSION TITLE
SyntaxError: Unexpected token < in JSON at position 0

### DISCUSSION THREAD

**Discussion by khashayarghajar** [vercel/next.js#50635] [Help] [answered]
### Summary

well, i'm trying to fetch data from an internal api but im facing this error `SyntaxError: Unexpected token < in JSON at position 0` on build time, but it works on development .

**Comment by icyJoseph** [comment]
What does "internal api" mean? Does it mean that during build time you are trying to fetch from `fetch("/api/some-route")`, and that route is a route of project? 

Or do you mean that it is some other API that is available in your network?

**Reply by khashayarghajar** [reply]
yes, im trying to fetch from `fetch("/api/some-route")`, and that route is a route of project

**Reply by icyJoseph** [reply, marked as answer]
That's the problem. During build time there's no server to respond to that request. In general, this is often a build crash or an anti-pattern.

You are already on the server, so why make a new request out to the internet and back to the server itself? During build time there's no server, and at runtime you pay with an extra, unnecessary request.

This part of the Page Dir docs https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props#getserversideprops-or-api-routes says it best:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/pages/building-your-application/routing/api-routes) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

So just call the data source from within `getStaticProps`, or `getServerSideProps` or from the Server Component. No need to make extra hoops.

For what is worth, Nuxt, a Vue based SSR framework, has a special `$fetch` function, which can skip the extra jump https://nuxt.com/docs/guide/concepts/server-engine#direct-api-calls

**Reply by khashayarghajar** [reply]
shukran

**Reply by Zenith2198** [reply]
I'm having the same exact issue. I am using Next.JS App Router, which [recommends you use fetch on the server](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-server-with-fetch). Is there any resolution, or is App Router incompatible with Vercel currently using this pattern?

**Reply by asumaran** [reply]
@Zenith2198 I got the same error, wondering what should we use instead? I'm using fetch to get data from an api route.

**Reply by choubari** [reply]
@Zenith2198 @asumaran I went through the same error today with App Router, and that was because I was fetching a **Route Handler** from a **Server Component**, while the docs recommends using route handlers (only) to fetch [on client components](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-route-handlers)...

In my code, I had a server component and I used an "extra layer" of fetching through a route handler _(which calls a third party library and not an api with fetch)_  just to be able to use `next: { revalidate: 0 }` or `cache: ''no-store"` as an argument on my first fetch at the server component level. While actually it is not necessary to add that extra layer with a route handler, I could just use what the docs recommend ([Fetching data on the Server with third-party libraries](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-server-with-third-party-libraries))

**Reply by asumaran** [reply]
@choubari in my case it wasn't working because I was fetching data from an endpoint in the same NextJS instance. I guess at build time the API routes aren't available. Fetching at another services worked just fine. I ended up fetching the database directly without the need of a HTTP request.
